src/tiled/objectsdock.cpp

```diff
@@ -454,3 +509,44 @@ void ObjectsView::synchronizeSelectedItems()
                              QItemSelectionModel::Clear);
     mSynching = false;
 }
+
+ObjectsFilterModel::ObjectsFilterModel(QObject *parent)
+    : QSortFilterProxyModel(parent)
+{
+}
+
+bool ObjectsFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
+{
+    // Show all if filter string is empty. Prevents hiding empty groups
+    if(filterRegExp().isEmpty())
+        return true;
+
+    QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
+
+    // sourceParent of a group has no model
+    if(!sourceParent.model())
```

Comment:
This is a somewhat interesting and not entirely correct check. There is no model, because the parent index is invalid, so a more sensible check may be `if (!sourceParent.isValid())`. However, either check won't work since layers now form a hierarchy and an object layer within a group layer _will_ have a valid parent, which would lead those object layers to be filtered as if they were an object.

Of course, group layers are a problem of their own as well, since whether they should be filtered or not would depend on whether any of their child layers are object groups that contain any non-filtered objects.

Fortunately, we know we're dealing with a `MapObjectModel` and I would suggest working through its interface directly instead of requesting data through `Qt::DisplayRole`. That way, you can use `toGroupLayer`, `toObjectGroup` and `toMapObject`. And to avoid having to convert between `ReversingProxyModel` indexes here, I would change the order in which the two proxies are applied so that the filtering happens before the reversing.